In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.datasets import load_iris
import matplotlib.pyplot as plt

## 1. Tutorial on Iris Dataset

In [3]:
# 1. Load the dataset
iris = load_iris()
X = iris.data
y = iris.target

# 2. Split the data into training and testing sets
# 70% training, 30% testing
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Create and train the KNN model with k=5
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)

# 4. Make predictions
y_pred = knn.predict(X_test)

# 5. Evaluate the model
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

# Optional: Classification report for precision, recall, f1-score
print("\nClassification Report:")
print(classification_report(y_test, y_pred))


Accuracy: 1.00

Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45



## 2. Apply KNN algorithm on the wine dataset and observe accuracy on training and test dataset with different values of ‘K’(2 to 5)

In [4]:
# Import required libraries
from sklearn.datasets import load_wine


# 1. Load the Wine dataset
wine = load_wine()
X = wine.data
y = wine.target

# 2. Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Train KNN for K=2 to 5 and evaluate accuracy
print("Wine Dataset - KNN Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)

    # Training accuracy
    y_train_pred = knn.predict(X_train)
    train_acc = accuracy_score(y_train, y_train_pred)

    # Test accuracy
    y_test_pred = knn.predict(X_test)
    test_acc = accuracy_score(y_test, y_test_pred)

    print(f"K={k} --> Training Accuracy: {train_acc:.2f}, Test Accuracy: {test_acc:.2f}")


Wine Dataset - KNN Accuracy:
K=2 --> Training Accuracy: 0.85, Test Accuracy: 0.70
K=3 --> Training Accuracy: 0.84, Test Accuracy: 0.74
K=4 --> Training Accuracy: 0.80, Test Accuracy: 0.72
K=5 --> Training Accuracy: 0.77, Test Accuracy: 0.74


## 3. Apply KNN algorithm on breast cancer dataset and observe accuracy on training and test datasetwith different values of ‘K’with different values of ‘K’(2 to 5)

In [5]:
# Import required libraries
from sklearn.datasets import load_breast_cancer


# 1. Load the Breast Cancer dataset
cancer = load_breast_cancer()
X = cancer.data
y = cancer.target

# 2. Split into training and testing sets (70% train, 30% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 3. Train KNN for K=2 to 5 and evaluate accuracy
print("Breast Cancer Dataset - KNN Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)

    # Training accuracy
    y_train_pred = knn.predict(X_train)
    train_acc = accuracy_score(y_train, y_train_pred)

    # Test accuracy
    y_test_pred = knn.predict(X_test)
    test_acc = accuracy_score(y_test, y_test_pred)

    print(f"K={k} --> Training Accuracy: {train_acc:.2f}, Test Accuracy: {test_acc:.2f}")


Breast Cancer Dataset - KNN Accuracy:
K=2 --> Training Accuracy: 0.96, Test Accuracy: 0.93
K=3 --> Training Accuracy: 0.95, Test Accuracy: 0.94
K=4 --> Training Accuracy: 0.94, Test Accuracy: 0.95
K=5 --> Training Accuracy: 0.92, Test Accuracy: 0.96


## 4. Apply KNN algorithm on thyroid dataset and observe accuracy on training and test datasetwith different values of ‘K’(2 to 5)

In [9]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [11]:
import pandas as pd

# Load the .data file (no headers in UCI datasets)
file_path = '/content/drive/MyDrive/Datsets/cleaned_dataset_Thyroid1.csv'
df = pd.read_csv(file_path)
print(df.info)
# Optional: Check first 5 rows
df.head()


<bound method DataFrame.info of        age  sex  on thyroxine  query on thyroxine  on antithyroid medication  \
0     41.0  0.0             0                   0                          0   
1     23.0  0.0             0                   0                          0   
2     46.0  1.0             0                   0                          0   
3     70.0  0.0             1                   0                          0   
4     70.0  0.0             0                   0                          0   
...    ...  ...           ...                 ...                        ...   
3766  30.0  0.0             0                   0                          0   
3767  68.0  0.0             0                   0                          0   
3768  74.0  0.0             0                   0                          0   
3769  72.0  1.0             0                   0                          0   
3770  64.0  0.0             0                   0                          0   

      s

,age,sex,on thyroxine,query on thyroxine,on antithyroid medication,sick,pregnant,thyroid surgery,I131 treatment,query hypothyroid,...,TSH measured,TSH,T3 measured,TT4 measured,TT4,T4U measured,T4U,FTI measured,FTI,binaryClass
0,41.0,0.0,0,0,0,0,0,0,0,0,...,1,1.30,1,1,125.0,1.0,1.140,1,109.000000,0
1,23.0,0.0,0,0,0,0,0,0,0,0,...,1,4.10,1,1,102.0,0.0,0.995,0,110.469649,0
2,46.0,1.0,0,0,0,0,0,0,0,0,...,1,0.98,0,1,109.0,1.0,0.910,1,120.000000,0
3,70.0,0.0,1,0,0,0,0,0,0,0,...,1,0.16,1,1,175.0,0.0,0.995,0,110.469649,0
4,70.0,0.0,0,0,0,0,0,0,0,0,...,1,0.72,1,1,61.0,1.0,0.870,1,70.000000,0


In [12]:
# 2. Ensure all columns are numeric (if needed)
df = df.apply(pd.to_numeric, errors='coerce')
df = df.dropna()  # drop any rows with NaN

# 3. Separate features and target
X = df.drop('binaryClass', axis=1).values
y = df['binaryClass'].values.astype(int)

# 4. Split into training (70%) and test (30%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 5. Train KNN for K=2 to 5 and observe accuracy
print("Thyroid Dataset - KNN Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)

    # Training and test accuracy
    train_acc = accuracy_score(y_train, knn.predict(X_train))
    test_acc = accuracy_score(y_test, knn.predict(X_test))

    print(f"K={k}: Training Accuracy={train_acc:.2f}, Test Accuracy={test_acc:.2f}")


Thyroid Dataset - KNN Accuracy:
K=2: Training Accuracy=0.97, Test Accuracy=0.95
K=3: Training Accuracy=0.97, Test Accuracy=0.95
K=4: Training Accuracy=0.96, Test Accuracy=0.95
K=5: Training Accuracy=0.96, Test Accuracy=0.96


## 5. Apply KNN algorithm on seed dataset and observe accuracy on training and test datasetwith different values of ‘K’(2 to 5)

In [17]:
import warnings
warnings.filterwarnings('ignore')


In [19]:
# 1. Load the .txt dataset
file_path = '/content/drive/MyDrive/Datsets/seeds_dataset.txt'  # dataset path
# The data is space/tab separated
seeds = pd.read_csv(file_path, sep='\s+', header=None)  # \s+ handles multiple spaces/tabs

# 2. Assign column names
seeds.columns = ['area','perimeter','compactness','length_of_kernel','width_of_kernel',
              'asymmetry_coefficient','length_of_groove','class']

# 3. Separate features and target
X = seeds.drop('class', axis=1).values
y = seeds['class'].values

# 4. Split into training (70%) and test (30%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# 5. Train KNN for K=2 to 5 and observe accuracy
print("Seed Dataset - KNN Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(X_train, y_train)

    train_acc = accuracy_score(y_train, knn.predict(X_train))
    test_acc = accuracy_score(y_test, knn.predict(X_test))

    print(f"K={k}: Training Accuracy={train_acc:.2f}, Test Accuracy={test_acc:.2f}")


Seed Dataset - KNN Accuracy:
K=2: Training Accuracy=0.97, Test Accuracy=0.89
K=3: Training Accuracy=0.94, Test Accuracy=0.87
K=4: Training Accuracy=0.94, Test Accuracy=0.89
K=5: Training Accuracy=0.92, Test Accuracy=0.89


## 6. Perform all above que(2 to 5) using 10 fold cross validation and observe accuracy on training and test datasetwith different values of ‘K’(2 to 5)

In [24]:
import warnings
warnings.filterwarnings('ignore')  # suppress warnings

import pandas as pd
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.datasets import load_iris, load_wine, load_breast_cancer






# Example with multiple datasets
datasets = {
    'Iris': load_iris(),
    'Wine': load_wine(),
    'Breast Cancer': load_breast_cancer(),

}



for name, dataset in datasets.items():
    X = dataset.data
    y = dataset.target

    print(f"\n{name} Dataset - 10-Fold Cross-Validation Accuracy:")
    for k in range(2, 6):
        knn = KNeighborsClassifier(n_neighbors=k)

        # 10-fold cross-validation
        scores = cross_val_score(knn, X, y, cv=10)  # default scoring='accuracy'
        mean_acc = scores.mean()

        print(f"K={k}: CV Accuracy={mean_acc:.2f}")



Iris Dataset - 10-Fold Cross-Validation Accuracy:
K=2: CV Accuracy=0.95
K=3: CV Accuracy=0.97
K=4: CV Accuracy=0.97
K=5: CV Accuracy=0.97

Wine Dataset - 10-Fold Cross-Validation Accuracy:
K=2: CV Accuracy=0.67
K=3: CV Accuracy=0.72
K=4: CV Accuracy=0.66
K=5: CV Accuracy=0.68

Breast Cancer Dataset - 10-Fold Cross-Validation Accuracy:
K=2: CV Accuracy=0.91
K=3: CV Accuracy=0.93
K=4: CV Accuracy=0.92
K=5: CV Accuracy=0.93


In [22]:
from sklearn.model_selection import cross_val_score, KFold

file_path = '/content/drive/MyDrive/Datsets/cleaned_dataset_Thyroid1.csv'
df= pd.read_csv(file_path)
# Ensure numeric values
df = df.apply(pd.to_numeric, errors='coerce').dropna()

# Features and target
X = df.drop('binaryClass', axis=1).values
y = df['binaryClass'].values.astype(int)

# 10-fold cross-validation
kf = KFold(n_splits=10, shuffle=True, random_state=42)

print("Thyroid Dataset - 10-Fold Cross-Validation Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, X, y, cv=kf)
    print(f"K={k}: CV Accuracy={scores.mean():.2f}")

Thyroid Dataset - 10-Fold Cross-Validation Accuracy:
K=2: CV Accuracy=0.95
K=3: CV Accuracy=0.95
K=4: CV Accuracy=0.95
K=5: CV Accuracy=0.95


In [26]:
# Load Seed dataset (assuming it's already a CSV or TXT loaded into pandas)
file_path = '/content/drive/MyDrive/Datsets/seeds_dataset.txt'
seeds = pd.read_csv(file_path, sep='\s+', header=None)
seeds.columns = ['area','perimeter','compactness','length_of_kernel','width_of_kernel',
                 'asymmetry_coefficient','length_of_groove','class']

# Separate features and target
X = seeds.drop('class', axis=1).values
y = seeds['class'].values

# 10-fold cross-validation
kf = KFold(n_splits=10, shuffle=True, random_state=42)

print("Seed Dataset - 10-Fold Cross-Validation Accuracy:")
for k in range(2, 6):
    knn = KNeighborsClassifier(n_neighbors=k)
    scores = cross_val_score(knn, X, y, cv=kf)
    print(f"K={k}: CV Accuracy={scores.mean():.2f}")

Seed Dataset - 10-Fold Cross-Validation Accuracy:
K=2: CV Accuracy=0.90
K=3: CV Accuracy=0.90
K=4: CV Accuracy=0.90
K=5: CV Accuracy=0.89


# Observation

The KNN algorithm shows consistently high accuracy across all datasets. For Iris and Breast Cancer, CV accuracy is above 90%, while Wine and Seed datasets have slightly lower accuracy (67–90%). Increasing K generally stabilizes performance, reducing overfitting seen at lower K. Thyroid dataset maintains 95% accuracy for all K. Overall, KNN performs well with small to moderate K and benefits from 10-fold cross-validation for reliable evaluation.



Using 10-fold cross-validation, KNN maintains strong performance across datasets. Iris and Breast Cancer achieve very high CV accuracies (91–97%), indicating robust generalization. Wine dataset shows moderate accuracy (66–72%), reflecting class overlap or complexity. Thyroid dataset remains consistent at 95%, and Seed dataset achieves 89–90%. Increasing K slightly smooths results, reducing variance but not drastically affecting overall accuracy. CV provides reliable performance estimation compared to simple train-test split.
